src/ert/com_protocol/_state_machine.py

```diff
@@ -0,0 +1,240 @@
+# pylint: disable=E1101 # might be fixed when we migrate to proto3
+import logging
+from functools import singledispatchmethod
+from typing import Optional, Union
+
+import ert.com_protocol._schema_pb2 as pb2
+
+logger = logging.getLogger(__name__)
+
+
+class _StateHandle:
+    """The :class:`_StateHandle` implements a mechanism for handling the protobuf
+    defined state machine.
+    General note:
+    We cannot directly assign values to a protobuf-map; eg:
+    experiment.ensembles[ens_id] = Ensemble(id=ens_id)
+    Useful links:
+    https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/struct
+    https://developers.google.com/protocol-buffers/docs/reference/python-generated#map-fields
+    """
+
+    def __init__(self) -> None:
+        self._experiment: Optional[pb2.Experiment] = None
+
+    def get_experiment(self, _id: pb2.ExperimentId) -> pb2.Experiment:
+        logger.debug(f"experiment id: {_id}")
+        if self._experiment:
+            if self._experiment.id.id != _id.id:
```

Comment:
Is this some kind of "authorization"? Why is the `_id` argument needed, we can't just return `self._experiment`?